### DISCUSSION TITLE
Keycloak Quarkus volatile installation data

### DISCUSSION THREAD

**Discussion by inglor** [keycloak/keycloak#10323] [Archive]
Hello,

I'm the maintainer of `keycloak` package in Arch Linux. With the new `17.x` version I'm looking into packaging keycloak under `/usr/share` which I'd argue is where it belongs. 

This proved challenging since the `kc.sh build` command tries to modify files under the `<keycloak-home>/lib/quarkus/` directory. According to [Linux FHS] the `/usr` supposed to be read-only and such data should reside in `/var/lib`.

So I was searching for a way to have the installation of `keycloak` in `/usr/share/java/keycloak` and the `quarkus` lib in `/var/lib`/keycloak` since it's volatile and can change depending of the configuration (e.g. change of db provider). Looking to the code of Quarkus it seems like it is serializing the configuration and persist it in file system (I think).

I also checked if I could move the quarkus data to `/var/lib/keycloak/` and have a symlink which will point to the /usr/share/keycloak/ but apparently this doesn't work either as they expect a directory explicitly.

So, is there a way to support the above structure? Can Keycloak with Quarkus be instructed to persist these volatile data elsewhere, maybe a config which I didn't find?

Thanks

[Linux FHS]: https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard

**Comment by DGuhr** [comment]
Hey @inglor. Thanks for reaching out. You are right, we're using the Quarkus mutable-jar approach and it's augmentation mechanism, where files are getting serialized, creating a new "closed world assumption" and the benefits in terms of startup/memory footprint based on that. You can read more about it here: https://quarkus.io/guides/reaugmentation 

cc @pedroigor / @stuartwdouglas - do you know perhaps if the location for the augmented directories & files is configurable?

**Comment by stuartwdouglas** [comment]
Maybe you could use a symlink? Looking at the code I think it would work, but I haven't tested it. 

At the moment the path structure is hard coded, but we could potentially look at changing that.

**Reply by inglor** [reply]
Hi @stuartwdouglas, thanks for the suggestion. That was the first thing I tried but the code expects a directory and fails with a symlink. Also I tried symlinking individual files but again the code overrides them with actual files when doing so.

If you can investigate what's possible with the configuration of the path it would be greatly appreciated!

**Reply by inglor** [reply]
> At the moment the path structure is hard coded, but we could potentially look at changing that.

Any update on this?

**Comment by mbert** [comment]
I have run into the same problem and done some research. If the `lib/quarkus` directory is a symlink, then the `build` command fails. The reason for this is how [quarkus](https://github.com/quarkusio/quarkus/) is implemented: it requires the `quarkus` directory to be a real directory. I am actually not sure whether this behaviour is a bug or a feature (because security plays some role here).

The most desirable solution to this problem (and I agree, `quarkus` _should_ reside somewhere under _/var/lib_) would be a configuration option to KeyCloak allowing to place `quarkus` (and presumably also `quarkus-run.jar`) somewhere else rather than keeping its location hard-coded as under `lib`, so that a symbolic link would not even be necessary.

However I have briefly taken a look at both the KC and quarkus code, and to me it has not become completely clear whether this can be made configurable at all:
- in the KC code I cannot find any place where quarkus is actually _configured_ to reside somewhere (so maybe this is a default in quarkus?)
- in the quarkus code I cannot find any hint that the default directory structure can be changed

All my attempts to move or symlink stuff around have failed. Hence for now it looks like we will have to live with a mutable `lib/quarkus` directory, which I do not like at all for the same reasons as mentioned by the OP.

**Reply by pedroigor** [reply]
I'm not sure if we can do this on our side. I also agree that recognizing symlink is not ideal from a security point of view. But perhaps an option.

@alexeykazakov Do you think it is reasonable to provide an option in Quarkus to read `lib` directory using a path other than what is expected from the default path structure?

**Reply by mbert** [reply]
Maybe, alternatively, make installation paths configurable?

**Reply by alexeykazakov** [reply]
Everything you says, @pedroigor, sounds reasonable to me! But I guess you wanted to ping someone else. Not me :)

**Reply by pedroigor** [reply]
@alexeykazakov Oops, sorry :) It was targeted to @aloubyansky.

**Comment by aloubyansky** [comment]
There seem to be two options how to address this issue:
1. Add support for custom locations for certain directories of the `fast-jar` and `mutable-jar` (built-in Quarkus) packages (would be implemented in Quarkus and configured at Keycloak build/installation time)
2. Come up with a Keycloak-specific Quarkus package type. Quarkus was designed to allow extensions to implement their own packaging types (would be implemented in the Keycloak Quarkus extension and configured at Keycloak build/installation time).

**Reply by inglor** [reply]
Glad I see some ideas/discussion to this one. Either would work but I think maybe the one which would be more contained within Keycloak would be 2, right?
Happy to take a (rough) stab on it if someone points to me the right direction :)

**Reply by pedroigor** [reply]
@inglor Yeah, looks like #2 is worth exploring. Perhaps having our own package type opens the door for more customizations on how we build our distribution too.

However, for this particular case, I would avoid additional complexity on our side and just use an option from the package config to change the directory. Sounds simpler but I'm not familiar with how to create a custom package type.

@aloubyansky Or #2 is simple enough to achieve?

**Reply by aloubyansky** [reply]
1 would be easier/quicker to implement for sure.
2 could be a longer term option, that would need some design and Keycloak installation use-case a analysis

**Reply by aloubyansky** [reply]
I don't think there is a chance though adding this in quarkus 2.13, unless you absolutely can't live w/o it.

**Reply by mbert** [reply]
> I don't think there is a chance though adding this in quarkus 2.13, unless you absolutely can't live w/o it.

You mean both, variant 1 and 2?

**Reply by aloubyansky** [reply]
I meant 1, since that would have to be done in Quarkus itself.
2 would be done in Keycloak (unless we hit an issue requiring an adjustment in Quarkus).

**Reply by inglor** [reply]
Is this something that is considered / implemented for 2.1.4 or 2.2 even?

**Reply by pedroigor** [reply]
I'm not sure if we are going to be able to work on it as our roadmap is quite aggressive this year with higher priorities.

**Comment by Smiche** [comment]
A workaround we're planning to use is to build a custom image and copy the lib/quarkus folder to another location in the container image. Then at runtime the `/opt/keycloak/lib/quarkus` folder is a volume and we modify the entrypoint to copy the read-only version of this folder we copied earlier. That way if we're updating to a newer version, the folder will have the latest files and we can still comply with read_only: true.

I wish there was a better solution.
